### PR TITLE
Add storage mining pool

### PR DESCRIPTION
**Pull request by mvines**
Storage needs a mining pool to pull rewards out of.

### CODE REVIEW

**Comment by mvines** [comment]
@sagar-solana - this is still a WIP, I'm gonna plumb up the wallet commands in this PR too.  Sharing early for FYI and any feedback though

**Comment by sagar-solana** [comment]
Yeah I think I see where you're going with this. Should be okay. 
Who all owns the mining pool keypair? Where does it get deposits from? 
I think we've answered those questions already but I wouldn't mind another explanation :angel:

**Comment by mvines** [comment]
> Who all owns the mining pool keypair?
Doesn't matter, they probably burn it actually.   You can't move funds out of the mining pool once contributed.

> Where does it get deposits from?
Manna from heaven.

**Comment by sagar-solana** [comment]
just fyi, proof validations will not work properly right now. I'm working on getting them fixed in time. In case you need to test something, you should be able to pass in a single `CheckedProof` in a proof Validation and that should work.

**Comment by sagar-solana** [comment]
Dunno why that test has started failing suddenly. :/

**Comment by mvines** [comment]
Oh are you seeing it as well?  I was assuming it was just this PR, and was gonna debug tomorrow

**Comment by sagar-solana** [comment]
It's likely something else. Replicators wait 50 seconds for the leader to make 16 rooted slots. Something changed recently that isn't making that happen and that test fails as a result of that.

File: programs/storage_api/src/storage_contract.rs [hunk 1]
```diff
@@ -456,7 +475,7 @@ mod tests {
 
         account.account.data.resize(4 * 1024, 0);
```
Comment (sagar-solana) [inline programs/storage_api/src/storage_contract.rs]: Maybe I missed this. Use the size const here?

File: programs/storage_api/src/storage_contract.rs [hunk 2]
```diff
@@ -69,6 +71,44 @@ impl<'a> StorageAccount<'a> {
...
+        let storage_contract = &mut self.account.state()?;
+        if let StorageContract::Uninitialized = storage_contract {
+            *storage_contract = StorageContract::MiningPool;
+            self.account.set_state(storage_contract)
+        } else {
+            Err(InstructionError::AccountAlreadyInitialized)?
+        }
+    }
+
+    pub fn initialize_replicator_storage(&mut self) -> Result<(), InstructionError> {
```
Comment (sagar-solana) [inline programs/storage_api/src/storage_contract.rs]: Nit: (especially given the readability benefits) this causes two writes to the account state the first time.

File: programs/storage_api/src/storage_instruction.rs [hunk 3]
```diff
@@ -27,12 +40,71 @@ pub enum StorageInstruction {
...
+        Instruction::new(
+            id(),
+            &StorageInstruction::InitializeValidatorStorage,
+            vec![AccountMeta::new(*storage_pubkey, false)],
+        ),
+    ]
 }
 
-pub fn mining_proof(
+pub fn create_replicator_storage_account(
```
Comment (sagar-solana) [inline programs/storage_api/src/storage_instruction.rs]: I'm less of a fan of this. It was cooler when it was implicitly figuring out what account to setup.

File: programs/storage_api/src/storage_contract.rs (hunk 2 above)
Comment (sagar-solana) [reply programs/storage_api/src/storage_contract.rs]: Nvm I didn't see the separate init instructions. Somehow it doesn't seem as elegant though.

File: programs/storage_api/src/storage_instruction.rs [hunk 4]
```diff
@@ -27,12 +40,71 @@ pub enum StorageInstruction {
...
+        Instruction::new(
+            id(),
+            &StorageInstruction::InitializeMiningPool,
+            vec![AccountMeta::new(*storage_pubkey, false)],
+        ),
+    ]
+}
+
+pub fn mining_proof(
+    storage_pubkey: &Pubkey,
```
Comment (rob-solana) [inline programs/storage_api/src/storage_instruction.rs]: does this TX need a from_pubkey for #3880 ?

File: programs/storage_api/src/storage_instruction.rs (hunk 3 above)
Comment (mvines) [reply programs/storage_api/src/storage_instruction.rs]: Yeah it's cooler but much more difficult to reason about, which makes it easier to create something exploitable.   Programs should be simple and dumb, never sneaky.

File: programs/storage_api/src/storage_instruction.rs (hunk 4 above)
Comment (mvines) [reply programs/storage_api/src/storage_instruction.rs]: I'm gonna make storage use a system account for fees immediately after I'm done with this PR

File: programs/storage_api/src/storage_instruction.rs (hunk 3 above)
Comment (sagar-solana) [reply programs/storage_api/src/storage_instruction.rs]: Well only certain instructions would implicitly setup the account. Others, like rewards, would return an error. Anyway what you have is totally fine. :)

Comment (mvines) [reply programs/storage_api/src/storage_instruction.rs]: IMO that puts too much faith in how the program handles an uninitialized account, so easy to break those implicit assumptions and create an exploit.  I like code that's dumber than me.

File: programs/storage_api/src/storage_contract.rs (hunk 1 above)
Comment (mvines) [reply programs/storage_api/src/storage_contract.rs]: ✔️

**Comment by mvines** [comment]
> It's likely something else. Replicators wait 50 seconds for the leader to make 16 rooted slots. 
> Something changed recently that isn't making that happen and that test fails as a result of that.
 
Oh man was stuff broken.  I don't understand how some of the local cluster tests were passing before

**Comment by sagar-solana** [comment]
> Oh man was stuff broken. 

Shows that something needs to be fixed in our tests. That silly "start_and_exit_with_config" or whatever is a sanity test for whether or not local_clusters with a non-default config even works. It shouldn't be the _only_ test to see some of these bugs haha

**Comment by mvines** [comment]
> adding the wallet command tests shows next guy how to fake it ;)

I see these commands making it into the testnet participation doc once replicators actually work.  Further adding to the already bloated `test_wallet_process_command()` and `test_wallet_parse_command()` tests seems like pretty low ROI.  Was there another place you were thinking of?

**Comment by mvines** [comment]
ffs https://www.githubstatus.com/incidents/nf4gcjtcsqdb

**Comment by sagar-solana** [comment]
:O Maybe I didn't build --all when I tried it? :\

**Comment by rob-solana** [comment]
> > adding the wallet command tests shows next guy how to fake it ;)
> 
> I see these commands making it into the testnet participation doc once replicators actually work. Further adding to the already bloated `test_wallet_process_command()` and `test_wallet_parse_command()` tests seems like pretty low ROI. Was there another place you were thinking of?

no, that's the place

your call

**Pull request merged**